### PR TITLE
Adding in RunAlways, run order, and ability to generate HTML Report

### PR DESCRIPTION
**Pull request by OctopusBob**
This pull request will add the following functionality to DbUp

## Adding in Script Types

Currently, all scripts are Run Once.  There are a lot of instances where it would be useful to have scripts which could always be run during a deployment.  For example, you want to ensure a specific role always exists.  Or you want to refresh all the views after a deployment so it gets any schema updates.  Or you want to run a stored procedure to generate some same data.  

A new enum called ScriptType has been added.  At this point, it has two types of scripts, RunOnce and RunAlways.  RunAlways scripts will always be run, even if there is an entry in the journal.  RunOnce is the default and will continue to function as before.  In the future, I plan on adding in "RunOnChanged." But that involved changing a lot more than I wanted to (plus I would need migration scripts to add that column to all the table types we support).

```csharp
var upgradeEngineBuilder = DeployChanges.To
                            .SqlDatabase(connectionString, null) //null or "" for default schema for user
                            .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly(), script => script.StartsWith("SampleApplication.Scripts."), new SqlScriptOptions { ScriptType = ScriptType.RunOnce })
                            .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly(), script => script.StartsWith("SampleApplication.RunAlways."), new SqlScriptOptions { ScriptType = ScriptType.RunAlways })
                            .LogToConsole();
```

The advantages of using a script type over say a null journal is these scripts will be recorded in the journal for each deployment and you combine RunOnce and RunAlways scripts together on the same deployment.

I added these as overloaded methods to the existing "WithScripts[xxx]" because I felt the script type should be defined when adding scripts from an assembly.  You shouldn't do it after the fact because you would have to write a regular expression or something to match on the file name.

## Run Group Order
By default, all scripts are run in alphabetical order.  This works great when we only had to worry about run once scripts.  With the addition of RunAlways scripts that adds a bit of complexity.  Perhaps you have a set of scripts to run before a deployment and a set of scripts to run afterward.  The run group order will take that into account and order by that first and then by alphabetical.

```csharp
var upgradeEngineBuilder = DeployChanges.To
    .SqlDatabase(connectionString, null) //null or "" for default schema for user
    .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly(), script => script.StartsWith("SampleApplication.PreDeployment."), new SqlScriptOptions { ScriptType = ScriptType.RunAlways, RunGroupOrder = 1})
    .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly(), script => script.StartsWith("SampleApplication.Scripts."), new SqlScriptOptions { ScriptType = ScriptType.RunOnce, RunGroupOrder = 2})
    .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly(), script => script.StartsWith("SampleApplication.PostDeployment."), new SqlScriptOptions { ScriptType = ScriptType.RunAlways, RunGroupOrder = 3})
    .LogToConsole();
```

Example:
- PostDeployment
    - 0001_InsertDefaultData.sql
- PreDeployment
   - 0001_CreateRolesIfNotExists.sql
- Scripts
   - 0001_AddTable.sql
   - 0002_CreateStoredProcedure.sql

The order those scripts would be run in are:
- 0001_CreateRolesIfNotExist.sql
- 0001_AddTable.sql
- 0002_CreateStoredProcedure.sql
- 0001_InsertDefaultData.sql

## HTML Reports
Two great features are provided by Octopus Deploy, artifacts and manual interventions.  HTML reports will allow people to generate HTML reports and upload them to Octopus Deploy as an artifact so an approver in a manual intervention can review the changes prior to deploying them.

This was added as an extension method to the upgrade engine.

```csharp
    var connectionString =
        args.FirstOrDefault()
        ?? "Server=(local)\\SqlExpress; Database=MyApp; Trusted_connection=true";

    var upgrader =
        DeployChanges.To
            .SqlDatabase(connectionString)
            .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly())
            .LogToConsole()
            .Build();

    // --generateReport is the name of the example argument.  You can call it anything
    if (args.Any(a => "--generateReport".Equals(a, StringComparison.InvariantCultureIgnoreCase))) 
    {
        upgrader.GenerateUpgradeHtmlReport("C:\\DeploymentLocation\\UpgradeReport.html");
    }
    else
    {
        var result = upgrader.PerformUpgrade();

        if (!result.Successful)
        {
            Console.ForegroundColor = ConsoleColor.Red;
            Console.WriteLine(result.Error);
            Console.ResetColor();
#if DEBUG
            Console.ReadLine();
#endif                
            return -1;
        }
    }


    Console.ForegroundColor = ConsoleColor.Green;
    Console.WriteLine("Success!");
    Console.ResetColor();
    return 0;
```

samplehtmlreport

### CODE REVIEW

**Comment by dazinator** [comment]
This looks very useful, especially the report, I was about to look at doing something like that myself!

My only thought really is about the `Run Order` api, it's a shame we can't just infer the order based on order or method calls via the API. i.e so you don't have to say `runOrder: 1` , `runOrder:2` etc. I would say can't we just increment that counter ourselves each time `WithScripts` is called, except I realise there can be cases where you want to include scripts from multiple sources / assemblies, but execute them all in one go / the same run Order so incrementing the run order automatically wouldn't be desirable in that scenario.

So with that in mind, I guess without some rethinking of the general API, having that parameter is the easiest path!

**Comment by OctopusBob** [comment]
Thank you!

I was thinking the same thing on the runOrder.  My initial version had it incrementally increasing a counter by one.  I'd rather it be up to the developer than trying to figure it out.  In addition, if you don't specify the run order then all the scripts are run in alphabetical order, which is the default behavior.  My goal with all these changes it to keep the default behavior the same.

Also, for some reason, this test keeps failing.  

image

I updated `dbup-tests/approvalfiles/dbup-core.approved.cs` to include the additional methods.  Anything else I need to do?  "It works on my machine" :)

**Comment by AdrianJSClark** [comment]
The exception that is failing those tests looks like something is misconfigured in AppVeyor:
`System.Exception : Could not find a diff program to use`

I'm a little concerned that we are adding three new features ("Run Always" script type, "Ordering" of groups of scripts, HTML Reports) in one huge PR. It makes it difficult to tease apart the items and consider their impact individually. I don't want to discourage this great work, and apologies for flying by and only having time right now to look over this PR on the surface, but is there any chance of having these three features as three separate PRs? It would certainly make it less imposing for me personally to find time to properly review things.

**Comment by OctopusBob** [comment]
The HTML report can be pulled into a separate PR.  That is an extension class with some doco around it.  

ScriptType and RunOrder are a little trickier to separate.  Both are new properties on SqlScript.  95% of the changes in the PR is adding the appropriate overloaded methods to set those properties and documentation.  Those two properties are only used in `DefaultScriptFilter.cs` and `UpgradeEngine.cs`.  I could separate them out, but then you'd have two large PRs instead changing almost the exact same set of files instead of one :)

File: docs/more-info/html-report.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+It is important to build trust in any automated process.  Typically, deployment tools have the ability to have approvers who will verify changes prior to being deployed.  A great to build trust with the approver is to have the ability to see which scripts are going to be run.  Some of these deployment systems have the ability to publish artifacts for approvers to review them.
```
Comment (jburger) [inline docs/more-info/html-report.md]: Sorry to nitpick - but there is a grammar error here: `A great to build trust`.

Totally agree with the sentiment here though!

**Comment by jburger** [comment]
FYI, I've merged https://github.com/DbUp/DbUp/pull/380 so if you pull that into this PR, your build should pass.

File: src/dbup-core/Engine/SqlScript.cs [hunk 2]
```diff
@@ -11,27 +12,55 @@ namespace DbUp.Engine
...
+
+        /// <summary>
+        /// Gets the type of script this is.
+        /// </summary>
+        public ScriptType ScriptType { get; }
+
+        /// <summary>
+        /// Gets run group order, by default all scripts run in alphabetical order, use this when you want to specify a group of scripts to run first or last
+        /// </summary>
+        public int runOrder { get; set; }
```
[the remark this answers is not in the document]
Comment (OctopusBob) [reply src/dbup-core/Engine/SqlScript.cs]: These properties over to a SqlScriptOptions class.

File: src/dbup-core/Helpers/UpgradeEngineHtmlReport.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    <meta http-equiv=""X-UA-Compatible"" content=""IE=edge"">                  
+    <meta http-equiv=""Content-Language"" content=""en"">
+    <link rel=""stylesheet"" href=""https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css"" integrity=""sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO"" crossorigin=""anonymous"">    
+    <title></title>               
+    <style> 
+    </style>
+</head>
+<body>
+    <noscript>
+            <div ""alert alert-danger"">& nbsp;JavaScript execution is currently disabled. Enable JavaScript on your browser to view the change report.</div>
```
[the remark this answers is not in the document]
Comment (OctopusBob) [reply src/dbup-core/Helpers/UpgradeEngineHtmlReport.cs]: Ahh, dang it, thanks for catching that.

File: src/dbup-tests/RunGroupOrderScenarios.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+using TestStack.BDDfy;
+using Xunit;
+
+namespace DbUp.Tests
+{
+    [Story(
+         AsA = "As a DbUp User",
+         IWant = "I want to DbUp to upgrade my database to the latest version using my own sort order",
+         SoThat = "So that my application's database is up to date")]
+    public class runOrderScenarios
```
[the remark this answers is not in the document]
Comment (OctopusBob) [reply src/dbup-tests/RunGroupOrderScenarios.cs]: The perils of find/replace.  I have updated the name to match what it is testing, RunGroupOrderScenarios.

File: docs/more-info/html-report.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (OctopusBob) [reply docs/more-info/html-report.md]: Thanks, I got that fixed :)

File: src/dbup-core/DbUpDefaults.cs [hunk 5]
```diff
@@ -5,5 +5,10 @@ namespace DbUp
     public static class DbUpDefaults
     {
         public static Encoding DefaultEncoding = Encoding.UTF8;
+
+        /// <summary>
+        /// The default order group everything is run in.  It is set to 100 to allow for scripts to run before the "standard" deployment as well as handle scripts to run after the deployment
+        /// </summary>
+        public static int DefaultRunOrder = 100;
```
[the remark this answers is not in the document]
Comment (OctopusBob) [reply src/dbup-core/DbUpDefaults.cs]: Agreed, that is a much better name.  Updated.

File: src/dbup-core/Engine/LazySqlScript.cs [hunk 6]
```diff
@@ -16,7 +17,31 @@ public class LazySqlScript : SqlScript
...
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LazySqlScript"/> class.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <param name="scriptType">The script type.</param>        
+        /// <param name="contentProvider">The delegate which creates the content at execution time.</param>
+        public LazySqlScript(string name, ScriptType scriptType, Func<string> contentProvider)
+            : this(name, scriptType, DbUpDefaults.DefaultRunOrder, null)
```
[the remark this answers is not in the document]
Comment (OctopusBob) [reply src/dbup-core/Engine/LazySqlScript.cs]: Good catch.  Thank you

**Comment by rupoase** [comment]
Hello everyone,
I'm having an issue when I'm trying to use the Filter on script.StartsWith(string) and RunAlways ScriptType for the WithScriptsEmbeddedInAssembly method.
What I want is always use the last version of one stored procedure. Also, I want to always log to the journal in order to know what was the actual version of the file (lookup in Git). The problem is that DbUp tells me that there isn't any new scripts to execute:
```shell
Beginning database upgrade
Checking whether journal table exists..
Fetching list of already executed scripts.
No new scripts need to be executed - completing.
```
If I remove the filter, the script is picked up and executed against the database again, and logged into the journal.
```csharp 
var result = DeployChanges.To
                .SqlDatabase(connectionString)
                .WithScriptsEmbeddedInAssembly(
                    Assembly.GetEntryAssembly(),
                    (string script) => script.StartsWith("script"),
                    new SqlScriptOptions
                    {
                        ScriptType = ScriptType.RunAlways
                    }
                )
                .LogToConsole()
                .Build();
```
I'm using .NetCore 1.6 with .NetStandard 1.6 but I'll migrate to NetStandard 2.0 in order to be able to use .NetCore 2.1+

**Pull request merged**